himembind.cpp

```diff
@@ -16,71 +16,188 @@
+     */
+    int i = 0;
+    while (i < call_name_len) {
+        if (call_name[i] != valid_name[i])
+            return 0;
+        i++;
+    }
+
+    return 1;
+}
+
+/**
+ * check valid character
+ * @param character
+ * @return
+ */
+int is_valid_character(char character) {
+    if (isalpha(character) || isnumber(character))
+        return 1;
+    /**
+     * check other allowed character '-','.','_'
+     */
+    if (character == '-' || character == '.' || character == '_')
+        return 1;
+
+    return 0;
+}
+
+/**
+ * check whether single argument is valid
+ * @param arg
+ * @return result
+ */
+int check_single_argument(const char *arg) {
+    int len = strlen(arg);
+    int i = 0;
+    while (i < len) {
+        if (!is_valid_character(arg[i]))
+            return 0;
+        i++;
+    }
+    return 1;
+}
+
+/**
+ * Check arguments are valid
+ * @param args arguments
+ * @param argc arguments count
+ * @return result
+ */
+int is_valid_arguments(char **args, int argc) {
+    int i = 0;
+    while (i < argc) {
+        if (!check_single_argument(args[i]))
+            return 0;
+        i++;
+    }
+    return 1;
+}
 
-int main(int argc, char** argv)
-{
-     int ret = 0;
-
-     const char* myname = argv[0];
-     ++argv;
-     --argc;
-
-     const char* callname = argv[0];
-
-     hwloc_topology_t topology;
-     unsigned long flags = 0;
-
-     hwloc_topology_init(&topology);
-     hwloc_topology_set_flags(topology, flags);
-     hwloc_topology_load(topology);
-
-     hwloc_cpuset_t cpuset = hwloc_bitmap_alloc();
-     hwloc_get_cpubind(topology, cpuset, HWLOC_CPUBIND_PROCESS);
-
-     char* ss = nullptr;
-     hwloc_bitmap_asprintf(&ss, cpuset);
-     fprintf(stderr, "cpuset: %s\n", ss);
-     free(ss);
-     ss = nullptr;
-
-     char* ss1 = nullptr;
-     hwloc_nodeset_t nodeset = hwloc_bitmap_alloc();
-     hwloc_cpuset_to_nodeset(topology, cpuset, nodeset);
-     hwloc_bitmap_asprintf(&ss1, nodeset);
-     fprintf(stderr, "nodeset: %s\n", ss1);
-     free(ss1);
-     ss1 = nullptr;
-
-     ret = hwloc_set_membind_nodeset(topology, nodeset, HWLOC_MEMBIND_BIND,
-                                     HWLOC_MEMBIND_STRICT);
-     if (ret) {
-          fprintf(stderr, "%s: Failed to set membind policy\n", myname);
-          goto exit_failure;
-     }
-
-     if (argc == 0) {
-          fprintf(stderr, "Usage: %s <cmd>\n", myname);
-          goto exit_failure;
-     }
-
-     ret = execvp(callname, argv);
-     if (ret) {
-          fprintf(stderr, "%s: Failed to launch executable \"%s\"\n", callname,
-                  argv[0]);
-          perror("execvp");
-          goto exit_failure;
-     }
-
-exit_failure:
-     hwloc_bitmap_free(cpuset);
-     hwloc_bitmap_free(nodeset);
-     hwloc_topology_destroy(topology);
-     return EXIT_FAILURE;
+/**
+ * Check call name is valid
+ * @param call_name
+ * @return result
+ */
+int check_call_name(const char *call_name) {
+    if (is_valid_call_name(call_name, "mpirun"))
```

Comment:
Sorry for late comment. I forgot to submit the review after commenting the code.

Reply:
I modified the code in the cpp style. You might check the latest code. Btw, as for the example you proposed, should python3 and python be added into valid call name (I obtained the example commands from @DechinPhy )? @yzotov

Reply:
What about if users want to specify a python that is not available on the PATH?
Something like this:

```bash
mpirun -np 2 himembind /home/user/bin/python shor.py
```

Reply:
What if we use some function like endsWith to handle such cases?

Reply:
That would be ok, although you would probably need to implement it yourself since `std::string::ends_with` is a C++20 feature.

Probably something along the lines of

```cpp
constexpr const char* valid_names[] = {
     "python",
     "python3"
};
const auto size = strlen(call_name);

for(const auto& name: valid_names) {
     const auto N = strlen(name);
     if (size >= N) {
          if (!std::equal(call_name + size - N, call_name + size,
                         name)) {
               return false;
          }
     }
}
return true;
```

This would naturally be somewhat easier with `std::string_view`:

```cpp
constexpr std::string_view valid_names[] = {
     "python",
     "python3"
};
for(const auto& name: valid_names) {
     if (size >= name.size()) {
          if (!std::equal(std::crbegin(call_name), std::crbegin(call_name) + name.size(),
                          std::crbegin(name)) {
               return false;
          }
     }
}
return true;
```

Reply:
Had to rework your solution a bit but I think now it is ok.